prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,1333 @@
+/*
+ * Copyright (c) 2016-2017 Cray Inc. All rights reserved.
+ * Copyright (c) 2017 Intel Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <fi_util.h>
+#include <inttypes.h>
+
+typedef unsigned long long int util_cache_entry_state_t;
+/* These are used for entry state and should be unique */
+#define UTIL_CES_INUSE       (1ULL << 8)    /* in use */
+#define UTIL_CES_STALE       (2ULL << 8)    /* cached for possible reuse */
+#define UTIL_CES_STATE_MASK  (0xFULL << 8)
+
+typedef unsigned long long int util_cache_entry_flag_t;
+/* One or more of these can be combined with the above */
+#define UTIL_CE_RETIRED     (1ULL << 61)   /* in use, but not to be reused */
+#define UTIL_CE_MERGED      (1ULL << 62)   /* merged entry, i.e., not
+					    * an original request from
+					    * fi_mr_reg */
+#define UTIL_CE_UNMAPPED    (1ULL << 63)   /* at least 1 page of the
+					    * entry has been unmapped
+					    * by the OS */
```

Comment:
These values are used by the registration monitoring system. 

If a page is unmapped, then the flag reflects that with CE_UNMAPPED. 

If a registration has been merged with another registration, then the flag reflects that with CE_MERGED. 

If a registration was originally mapped to some region, but then the region was coalesced into a different, larger registration, the original registration would be marked as retired -- putting it on a short list to be directly free'd at a later time when the reference count would drop to 0. The point of marking it as retired is so that the registration system doesn't attempt to reuse the retired entry in lieu of the newer merged entry which encompasses the prior region + some new region.